Lib/string.py

```diff
@@ -141,6 +141,16 @@ def convert(mo):
                              self.pattern)
         return self.pattern.sub(convert, self.template)
 
+    def get_identifiers(self, *, raise_on_invalid=True):
```

Comment:
Two questions about the signature.  Maybe no-raise should be the default?   If so then what do you think about `safe=True` as the argument to control this?  It's a nod to `safe_substitute()`.  Unfortunately, because of the `**kws` API for `substitute()` we couldn't do it that way over there (although `mapping` currently breaks that assumption).

Reply:
I guess the question for me is, a developer may want to know two things about a template, and I'm aiming for this PR to solve for both: 1) is it valid? 2) what are the identifiers in it? Does it make sense for these two things to be learned through the same call? If so, should the default be to find out about (1) if you are asking (2)? If not, is `get_identifiers(safe=False)` an obvious way to check the validity of the template?

Reply:
Actually, that's an excellent point.  Should those two questions be answered by a single, new method?  It's useful to understand why we have `safe_substitute()` in the first place.  The original motivating use case (indeed for [PEP 292](https://www.python.org/dev/peps/pep-0292/) in general) was for internationalizing Python applications, such as GNU Mailman.  One thing you definitely do *not* want in that use case is for typos in translated strings to cause the server to crash.  it's better to have a broken translation.

Let me throw this idea out there:

* What if you added an `is_valid` read-only property that returned `True` or `False` depending on the validity of the template?
* Change `get_identifiers()` to a read-only property `identifiers` which returns the list as `get_identifiers()` does now.  Then the question is whether to throw an exception or not when the template is invalid.  Maybe it just silently ignores invalid identifiers?

Reply:
> Maybe get_identifiers() with no arguments (always safe), and is_valid() as a method?

Works for me!

Reply:
Last question: both `substitute` and `safe_substitute` raise `ValueError` if the pattern has an unexpected named group that matches. For an otherwise-valid template, should `is_valid()` return `True`, `False`, or raise `ValueError`? I'd suggest the latter, as it's a problem with the pattern, not the template (which is why `safe_substitute` raises an error).

Reply:
I went with raising an error for an invalid pattern (in both `is_valid` and `get_identifiers`), let me know if that works.

Reply:
Oops, I just saw this, which is why I asked my questions below.  I think you're right that a problem with the template should raise exceptions.  I'll resolve those comments below.

Reply:
The `ValueError` raised when the pattern is invalid is not currently mentioned for `substitute` and `safe_substitute`, should I add it to the docs for all the methods?

Reply:
I added documentation on the `ValueError` raised for an invalid pattern in the section on setting your own pattern (it can't happen if you're setting the pattern components).